terratorch/models/smp_model_factory.py

```diff
@@ -1,83 +1,387 @@
+            raise ValueError(msg)
+
+        backbone_kwargs = _extract_prefix_keys(kwargs, "backbone_")  # Encoder params should be prefixed backbone_
+        smp_kwargs = _extract_prefix_keys(backbone_kwargs, "smp_")  # Smp model params should be prefixed smp_
+        aux_params = _extract_prefix_keys(backbone_kwargs, "aux_")  # Auxiliary head params should be prefixed aux_
+        aux_params = None if aux_params == {} else aux_params
+
         if isinstance(pretrained, bool):
             if pretrained:
                 pretrained = "imagenet"
             else:
                 pretrained = None
-        if decoder == "unet":
-            model = smp.Unet(
-                encoder_name=backbone, encoder_weights=pretrained, in_channels=in_channels, classes=num_classes
-            )
+
+        # If encoder not currently supported by SMP (custom encoder).
+        if backbone not in smp_encoders:
+            # These params must be included in the config file with appropriate prefix.
+            required_params = {
+                "encoder_depth": smp_kwargs,
+                "out_channels": backbone_kwargs,
+                "output_stride": backbone_kwargs,
+            }
+
+            for param, config_dict in required_params.items():
+                if param not in config_dict:
+                    msg = f"Config must include the '{param}' parameter"
+                    raise ValueError(msg)
+
+            # Using new encoder.
+            backbone_class = self._make_smp_encoder(backbone)
+            backbone_kwargs["prepare_features_for_image_model"] = prepare_features_for_image_model
+            # Registering custom encoder into SMP.
+            _register_custom_encoder(backbone_class, backbone_kwargs, pretrained)
+
+            model_args = {
+                "encoder_name": "SMPEncoderWrapperWithPFFIM",
+                "encoder_weights": pretrained,
+                "in_channels": in_channels,
+                "classes": num_classes,
+                **smp_kwargs,
+            }
+        # Using SMP encoder.
         else:
-            msg = "Only unet decoder implemented"
-            raise NotImplementedError(msg)
+            model_args = {
+                "encoder_name": backbone,
+                "encoder_weights": pretrained,
+                "in_channels": in_channels,
+                "classes": num_classes,
+                **smp_kwargs,
+            }
+
+        model = model_module(**model_args, aux_params=aux_params)
+
         return SMPModelWrapper(
             model, relu=task == "regression" and regression_relu, squeeze_single_class=task == "regression"
         )
 
 
-class SMPModelWrapper(Model, nn.Module):
-    def __init__(self, smp_model, relu=False, squeeze_single_class=False) -> None:
-        super().__init__()
-        self.smp_model = smp_model
-        self.final_act = nn.ReLU() if relu else nn.Identity()
-        self.squeeze_single_class = squeeze_single_class
+def get_smp_decoder(
```

Comment:
this function and the class `SMPDecoderForPrithviWrapper` seem to only be used for creating decoders which match the format required by `PrithviModelFactory` right? Then, would it make sense to instead move them to the `PrithviModelFactory` class?

Reply:
This function is also used by _get_smp_encoder to create the dummy smp encoder.